### PR TITLE
[DSK-27] Users - read & delete endpoints

### PR DESCRIPTION
**Pull request by jan-zabloudil**
~~Supabase go client was also updated to provide new methods:~~
~~https://github.com/jan-zabloudil/supabase-go/pull/2~~

### CODE REVIEW

File: repository/utils/utils.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+			return fmt.Errorf("%w: %s", reperr.ErrUserAuthenticationFailed, err.Error())
+		case 404:
+			return fmt.Errorf("%w: %s", reperr.ErrUserAuthenticationFailed, err.Error())
+		}
+	}
+
+	return fmt.Errorf("%w: %s", reperr.GeneralErr, err.Error())
+}
+
+func WrapSupabaseAdminErr(err error) error {
```
Comment (CermakM) [inline repository/utils/utils.go:47]: I don't understand the name of this function. Why is this `AdminErr`? I get DBErr and AuthErr from the previous functions, but here it's not immediately apparent to me.

File: repository/model/users.go
Comment (CermakM) [inline repository/model/users.go:1]: Singular `user.go`.

File: supabase/config.toml
Comment (CermakM) [inline supabase/config.toml:1]: Singular `user.go`.

**Review by CermakM — CHANGES_REQUESTED** [review]

File: repository/utils/utils.go (hunk 1 above)
Comment (jan-zabloudil) [reply repository/utils/utils.go:52]: Auth endpoint is called in order to authenticate user. User’s bearer token is provided in the authorization header.
**Error responses are following:** 

- 401 = invalid bearer token -> ReleaseManager returns 401
- 404 = user not found (valid token but user no longer exists in the database) -> ReleaseManager returns 401.

---
Admin endpoins are called in order to retrive users. Supabase secret key must be provided to access auth data.
**Error responses are following:**
- 401 = cannot access resource (invalid supabase secret key provided or only supabase anon key provided) -> ReleaseManager returns 500 (cannot access db because bad credentials were provided in the config)
- 404 = user not found (no user found for provided uuid) -> ReleaseManager returns 404

I needed to distinquish between those cases to return correct response. 
And I find better to create separate function for Auth and Admin than checking error text as well. I find it better to rely only on returned status code.

**Comment by jan-zabloudil** [comment]
All review comments were addressed, lot of changes were made because i switched to `go.strv.io/net`.

This branch is rebased onto #12. So once this PR is approved, I am going to merge both branches, there is no point of doing review for #12 separately.

Sorry that commits are squashed already.

File: api-doc.yaml [hunk 2]
```diff
@@ -60,7 +60,7 @@ paths:
               schema:
                 $ref: '#/components/schemas/OrganizationSettings'
         '400':
-          $ref: '#/components/responses/ForbiddenErrorResponse'
+          $ref: '#/components/responses/BadRequestErrorResponse'
         '401':
```
Comment (jan-zabloudil) [inline api-doc.yaml:64]: Not related to this PR. Just fix of error I noticed in openapi.

File: repository/model/user.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import (
+	"time"
+
+	svcmodel "release-manager/service/model"
+
+	"github.com/google/uuid"
+	"github.com/nedpals/supabase-go"
+)
+
+func ToSvcUser(
```
Comment (CermakM) [inline repository/model/user.go:12]: Use 1 line per argument or all args in a line.

File: repository/model/user.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+	id, email string,
+	isAdmin, name, picture interface{},
```
Comment (CermakM) [inline repository/model/user.go:14]:
```suggestion
	isAdmin, name, picture any,
```

File: transport/middleware.go [hunk 5]
```diff
@@ -1,19 +1,80 @@
...
 
-func (h *Handler) recoverPanic(next http.Handler) http.Handler {
+func (h *Handler) auth(next http.Handler) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		defer func() {
-			if err := recover(); err != nil {
-				w.Header().Set("Connection", "close")
-				WriteServerErrorResponse(w, r, fmt.Errorf("%s", err))
+
+		w.Header().Add("Vary", "Authorization")
```
Comment (CermakM) [inline transport/middleware.go:18]: I believe you can get rid of Vary, browsers don't really support it and cache control happens based on Cache-Control or status response.

File: service/model/user.go [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+	ReadAll(ctx context.Context) ([]User, error)
+	Delete(ctx context.Context, id uuid.UUID) error
+}
+
+type User struct {
+	ID        uuid.UUID
+	Email     string
+	Name      string
+	AvatarUrl string
+	IsAdmin   bool
```
Comment (CermakM) [inline service/model/user.go:22]: I'd prefer `UserRole` custom type here, it's more flexible, but you can do that in a separate PR.

File: cmd/api/main.go [hunk 7]
```diff
@@ -2,85 +2,51 @@ package main
...
+	timex "go.strv.io/time"
+)
 
 func main() {
-	cfg := loadConfig()
+	ctx := context.Background()
+	cfg := loadConfig(ctx)
 
-	logger := slog.New(slog.NewTextHandler(os.Stdout, &slog.HandlerOptions{Level: cfg.LogLevel}))
+	logger := slog.New(slog.NewJSONHandler(os.Stdout, &slog.HandlerOptions{Level: cfg.LogLevel}))
```
Comment (CermakM) [inline cmd/api/main.go:24]: Logs are usually logged to Stderr

File: repository/errors/errors.go [hunk 8]
```diff
@@ -0,0 +1,12 @@
+package errors
+
+import "errors"
+
+var (
+	UnknownErr                                        = errors.New("supabase: unknown error")
+	ErrEntityNotExists                                = errors.New("supabase: entity does not exist")
+	ErrUserAuthenticationFailed                       = errors.New("supabase: user authentication by bearer token unsuccessful")
+	ErrAccessToResourceDenied                         = errors.New("supabsse: access to resource denied")
```
Comment (CermakM) [inline repository/errors/errors.go:9]:
```suggestion
	ErrAccessToResourceDenied                         = errors.New("supabase: access to resource denied")
```

File: transport/context.go
Comment (CermakM) [inline transport/context.go:1]: I would put this file until transport/util

**Review by CermakM — CHANGES_REQUESTED** [review]

File: service/model/user.go [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+var AnonUser = &User{}
+
+type User struct {
+	ID        uuid.UUID
+	Email     string
+	Name      string
+	AvatarUrl string
+	Role      UserRole
```
Comment (jan-zabloudil) [inline service/model/user.go:24]: UserRole type added (see /services/model/user_role).

Also all user methods such as isAdmin, isAnon were moved into service layer.

File: service/model/user_test.go
Comment (jan-zabloudil) [inline service/model/user_test.go:1]: Tests added for user and user_role

File: transport/middleware.go [hunk 10]
```diff
@@ -1,19 +1,79 @@
...
+				WriteInvalidAuthenticationResponse(w, err)
+				return
+			default:
+				WriteServerErrorResponse(w, err)
+				return
 			}
-		}()
+		}
+
+		r = utils.ContextSetUser(r, &user)
```
Comment (jan-zabloudil) [inline transport/middleware.go:47]: The service User object is now stored in the context.
I believe that storing the User struct in the context is acceptable, the object is not too big.

**Review by CermakM — APPROVED** [review]
Good job! Let's merge this. 💪
